### ISSUE TITLE
Provide Azure Service Bus plugin for adding correlation info to messages

### ISSUE THREAD

**Issue by tomkerkhove** [arcus-azure/arcus.messaging#22]
Provide Azure Service Bus plugin for adding correlation info to outbound messages based on the current `ICorrelationInfo`.

For an example of what a Service Bus plugin is, see [SeanFeldman/ServiceBus.AttachmentPlugin](https://github.com/SeanFeldman/ServiceBus.AttachmentPlugin).

**Comment by stijnmoreels** [comment]
I think we can use the ServiceBus plugin functionality to add correlation information:

```csharp
/// <summary>
    /// Azure ServiceBus plugin to add correlation information to the outbound message.
    /// </summary>
    public class CorrelationInfoServiceBusPlugin : ServiceBusPlugin
    {
        private readonly CorrelationInfo _correlationInfo;

        /// <summary>
        /// Initializes a new instance of the <see cref="CorrelationInfoServiceBusPlugin"/> class.
        /// </summary>
        public CorrelationInfoServiceBusPlugin(CorrelationInfo correlationInfo)
        {
            Guard.NotNull(correlationInfo, nameof(correlationInfo));

            _correlationInfo = correlationInfo;
        }

        /// <inheritdoc />
        public override string Name => nameof(CorrelationInfoServiceBusPlugin);

        /// <inheritdoc />
        public override Task<Message> BeforeMessageSend(Message message)
        {
            message.CorrelationId = _correlationInfo.OperationId;

            if (!String.IsNullOrWhiteSpace(_correlationInfo.TransactionId))
            {
                message.UserProperties.Add(PropertyNames.TransactionId, _correlationInfo.TransactionId);
            }

            return Task.FromResult(message);
        }
    }
```

**Comment by stijnmoreels** [comment]
Can we use auto DI container instead registering manually.

**Comment by stijnmoreels** [comment]
Is this what we look for:
https://github.com/SeanFeldman/ServiceBus.AttachmentPlugin/blob/develop/src/ServiceBus.AttachmentPlugin/AzureStorageAttachmentExtensions.cs

**Comment by tomkerkhove** [comment]
> Is this what we look for:
> [SeanFeldman/ServiceBus.AttachmentPlugin:src/ServiceBus.AttachmentPlugin/AzureStorageAttachmentExtensions.cs@`develop`](https://github.com/SeanFeldman/ServiceBus.AttachmentPlugin/blob/develop/src/ServiceBus.AttachmentPlugin/AzureStorageAttachmentExtensions.cs) [](/SeanFeldman/ServiceBus.AttachmentPlugin/blob/HEAD@{2020-01-06T08:45:08Z}/src/ServiceBus.AttachmentPlugin/AzureStorageAttachmentExtensions.cs)

Not really as this does the same as above, while we have to get it at runtime

**Comment by stijnmoreels** [comment]
Didn't we agree to not do this bc it's not possible with the current setup?

**Comment by tomkerkhove** [comment]
Can you provide docs for creating Service Bus Messages that covers this a bit?

Think it's called `AsServiceBusMessage` but I'm not sure

**Comment by stijnmoreels** [comment]
> Can you provide docs for creating Service Bus Messages that covers this a bit?
> 
> Think it's called `AsServiceBusMessage` but I'm not sure

I also guess you're not gonna be a fan of creating a separate factory or model that implicitly can be cast to a Service Bus message.

**Comment by tomkerkhove** [comment]
While that is another approach, I think we should keep it simple and do everything folks would need as a minimum.

I personally thing an extension on the message content that you want to send is simplest approach rather than having to know there is a builder.

Thoughts @SamVanhoutte @gverstraete @fgheysels? (see #72 for example of proposal for builder)

**Comment by stijnmoreels** [comment]
In the meantime...
![Alt Text](https://media.giphy.com/media/QBd2kLB5qDmysEXre9/giphy.gif)

**Comment by tomkerkhove** [comment]
Any input @gverstraete @fgheysels?

**Comment by gverstraete** [comment]
Not sure about the added value to be honest. It might be that I am missing context on this topic, that's why I didn't reply before.

For me, based on the initial question, The correlation id is more of a header thing than a body parameter... + changing the body for this is not always a good idea. Or do you want to enforce this with cloudevents?

**Comment by tomkerkhove** [comment]
Then I'm leaning more towards leaving it out for now but will wait for @fgheysels

**Comment by tomkerkhove** [comment]
@stijnmoreels this was closed in #82 but #72 is still waiting on final agreement.

Maybe it's best to hold this off for now and start with the extension and add the builder when there is demand for it - What do you think?

If so, feel free to ship Messaging 1.0

**Comment by stijnmoreels** [comment]
> @stijnmoreels this was closed in #82 but #72 is still waiting on final agreement.
> 
> Maybe it's best to hold this off for now and start with the extension and add the builder when there is demand for it - What do you think?
> 
> If so, feel free to ship Messaging 1.0

Yes, I agree. I'll first add some docs for the message pump custom deserialization first?

**Comment by tomkerkhove** [comment]
That's fine for me as well, thanks!
